server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndexQuery.java

```diff
@@ -51,16 +52,51 @@ public String getQuery() {
     return query;
   }
 
-  /**
-   * The number of search hits to return per Qualifier. Defaults to <tt>10</tt>.
-   */
-  public ComponentIndexQuery setLimit(int limit) {
-    checkArgument(limit >= 1, "Limit has to be strictly positive: %s", limit);
-    this.limit = Optional.of(limit);
-    return this;
+  public Set<String> getRecentlyBrowsed() {
+    return recentlyBrowsed;
   }
 
   public Optional<Integer> getLimit() {
-    return limit;
+    return Optional.ofNullable(limit);
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private String query;
+    private Collection<String> qualifiers = Collections.emptyList();
+    private Set<String> recentlyBrowsed = Collections.emptySet();
+    private Integer limit;
+
+    private Builder() {
+    }
+
+    public Builder setQuery(String query) {
+      checkArgument(query.length() >= 2, "Query must be at least two characters long: %s", query);
+      this.query = query;
+      return this;
+    }
+
+    public Builder setQualifiers(Collection<String> qualifiers) {
+      this.qualifiers = Collections.unmodifiableCollection(qualifiers);
+      return this;
+    }
+
+    public Builder setRecentlyBrowsed(Set<String> recentlyBrowsed) {
+      this.recentlyBrowsed = Collections.unmodifiableSet(recentlyBrowsed);
+      return this;
+    }
+
+    public Builder setLimit(@Nullable Integer limit) {
+      checkArgument(limit == null || limit > 0, "Limit has to be strictly positive: %s", limit);
```

Comment:
I do not think this is required for the moment, because this is not exposed to the user.